### PR TITLE
[Merged by Bors] - feat (LinearAlgebra/TensorProduct/RightExactness): right exactness of tensor product for algebras

### PR DESCRIPTION
**Pull request by AntoineChambert-Loir**
Prove right exactness of tensor product for algebras.

* `Algebra.TensorProduct.ker_map` computes the kernel of `Algebra.TensorProduct.map f g`

* `Algebra.TensorProduct.lTensor_ker` and `Algebra.TensorProduct.rTensor_ker` compute the kernels
  of `Algebra.TensorProduct.map f id` and `Algebra.TensorProduct.map id g`

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/LinearAlgebra/TensorProduct/RightExactness.lean [hunk 1]
```diff
@@ -360,3 +366,237 @@ theorem TensorProduct.map_ker :
...
+
+variable
+    {R : Type*} [CommSemiring R]
+    {A B : Type*} [Semiring A] [Semiring B] [Algebra R A] [Algebra R B]
+
+/- The following lemmas could be easier if `I ⊗[R] B` was naturally an `A ⊗[R] B` module,
+  and the map to `A ⊗[R] B` was known to be linear.
+  This depends on the B-module structure on a tensor product
+  whose use rapidly conflicts with
+  everything… -/
```
[the remark this answers is not in the document]
Comment (AntoineChambert-Loir) [reply Mathlib/LinearAlgebra/TensorProduct/RightExactness.lean]: I updated the documentation.

File: Mathlib/LinearAlgebra/TensorProduct/RightExactness.lean [hunk 2]
```diff
@@ -360,3 +366,237 @@ theorem TensorProduct.map_ker :
...
+        apply Ideal.mul_mem_left
+        apply Ideal.mem_map_of_mem
+        exact Submodule.coe_mem b
+        simp only [Submodule.coe_restrictScalars, Algebra.TensorProduct.tmul_mul_tmul,
+          mul_one, one_mul]
+    | add x y hx hy =>
+        rw [map_add]
+        apply Submodule.add_mem _ hx hy
+
+/-- The ideal of `A ⊗[R] B` generated by `I` is the image of `A ⊗[R] I` (carrier) -/
```
[the remark this answers is not in the document]
Comment (AntoineChambert-Loir) [reply Mathlib/LinearAlgebra/TensorProduct/RightExactness.lean]: I could eliminate carrier in benefit of restrictScalars. That removes two lemmas that did almost nothing (a rewrite and a refl)

File: Mathlib/LinearAlgebra/TensorProduct/RightExactness.lean [hunk 3]
```diff
@@ -9,10 +9,14 @@ import Mathlib.RingTheory.TensorProduct
 
 /-! # Right-exactness properties of tensor product
 
-* `TensorProduct.rTensor.surjective` asserts that one tensors
+## Modules
+
+## Modules
```
Comment (riccardobrasca) [inline Mathlib/LinearAlgebra/TensorProduct/RightExactness.lean:14]:
```suggestion
## Modules
```

File: Mathlib/LinearAlgebra/TensorProduct/RightExactness.lean [hunk 4]
```diff
@@ -361,3 +392,210 @@ theorem TensorProduct.map_ker :
...
+    RingHom.ker (map f (AlgHom.id R C)) =
+      (RingHom.ker f).map (Algebra.TensorProduct.includeLeft : A →ₐ[R] A ⊗[R] C) := by
+  rw [← Submodule.restrictScalars_inj R]
+  have : (RingHom.ker (map f (AlgHom.id R C))).restrictScalars R =
+    LinearMap.ker (LinearMap.rTensor C (AlgHom.toLinearMap f)) := rfl
+  rw [this, Ideal.map_includeLeft_eq]
+  rw [(rTensor_exact C f.toLinearMap.exact_subtype_ker_map hf).linearMap_ker_eq]
+  rfl
+
+lemma AlgHom.coe_ker {R A B : Type*} [CommSemiring R] [Semiring A] [Semiring B]
```
Comment (riccardobrasca) [inline Mathlib/LinearAlgebra/TensorProduct/RightExactness.lean:570]: Here and in the following lemmas you can use the variables defined above, right?

**Review by riccardobrasca — COMMENTED** [review]
Thanks!

bors d+

File: Mathlib/LinearAlgebra/TensorProduct/RightExactness.lean (hunk 4 above)
Comment (AntoineChambert-Loir) [reply Mathlib/LinearAlgebra/TensorProduct/RightExactness.lean:568]: yes, and while we're there, why not transferring this lemma to the adequate file?

Comment (AntoineChambert-Loir) [reply Mathlib/LinearAlgebra/TensorProduct/RightExactness.lean:568]: in fact, not quite : in that lemma, R is a CommSemiring, not a CommRing, and similarly for A, B.

Comment (riccardobrasca) [reply Mathlib/LinearAlgebra/TensorProduct/RightExactness.lean:568]: Ah yes, of course! Thanks.

Comment (AntoineChambert-Loir) [reply Mathlib/LinearAlgebra/TensorProduct/RightExactness.lean:568]: I moved it to the end of `RingTheory.Ideal.Operations`, where docs#find_home told me I could put it. It's where docs#RingHom.ker is defined.

**Pull request closed without merging**